src/vm/clrtocomcall.cpp

```diff
@@ -629,6 +631,342 @@ UINT32 CLRToCOMEventCallWorker(ComPlusMethodFrame* pFrame, ComPlusCallMethodDesc
+        OBJECTREF RetValType;
+        OBJECTREF RetVal;
+    } gc;
+    ZeroMemory(&gc, sizeof(gc));
+    GCPROTECT_BEGIN(gc);
+    {
+        // Retrieve the exposed type object for the interface.
+        gc.ItfTypeObj = pItfMT->GetManagedClassObject();
+
+        // Retrieve the name of the target member. If the member
+        // has a DISPID then use that to optimize the invoke.
+        DISPID dispId = DISPID_UNKNOWN;
+        hr = pItfMD->GetMDImport()->GetDispIdOfMemberDef(tkMember, (ULONG*)&dispId);
+        if (hr == S_OK)
+        {
+            WCHAR strTmp[ARRAYSIZE(W("4294967295"))];
+            _snwprintf_s(strTmp, COUNTOF(strTmp), _TRUNCATE, DISPID_NAME_FORMAT_STRING, dispId);
+            gc.MemberName = StringObject::NewString(strTmp);
+        }
+        else
+        {
+            gc.MemberName = StringObject::NewString(strMemberName);
+        }
+
+        CallsiteDetails callsite = CreateCallsiteDetails(pFrame);
+
+        // Arguments
+        CallsiteInspect::GetCallsiteArgs(callsite, &gc.Args, &gc.ArgsIsByRef, &gc.ArgsTypes);
+
+        // If call requires object wrapping, set up the array of wrapper types.
+        if (pMD->RequiresArgumentWrapping())
+            gc.ArgsWrapperTypes = SetUpWrapperInfo(pItfMD);
+
+        // Return type
+        TypeHandle retValHandle = callsite.MetaSig.GetRetTypeHandleThrowing();
+        gc.RetValType = retValHandle.GetManagedClassObject();
+
+        // the return value is written into the Frame's neginfo, so we don't
+        // need to return it directly. We can just have the stub do that work.
+        // However, the stub needs to know what type of FP return this is, if
+        // any, so we return the return size info as the return value.
+        if (callsite.MetaSig.HasFPReturn())
+        {
+           callsite.MetaSig.Reset();
+           ArgIterator argit{ &callsite.MetaSig };
+           fpRetSize = argit.GetFPReturnSize();
+           _ASSERTE(fpRetSize > 0);
+        }
+
+        // Create a call site for the invoke
+        MethodDescCallSite forwardCallToInvoke(METHOD__CLASS__FORWARD_CALL_TO_INVOKE, &gc.ItfTypeObj);
+
+        // Prepare the arguments that will be passed to the method.
+        ARG_SLOT invokeArgs[] =
+        {
+            ObjToArgSlot(gc.ItfTypeObj),
+            ObjToArgSlot(gc.MemberName),
+            (ARG_SLOT)binderFlags,
+            ObjToArgSlot(pFrame->GetThis()),
+            ObjToArgSlot(gc.Args),
+            ObjToArgSlot(gc.ArgsIsByRef),
+            ObjToArgSlot(gc.ArgsWrapperTypes),
+            ObjToArgSlot(gc.ArgsTypes),
+            ObjToArgSlot(gc.RetValType)
+        };
+
+        // Slot for return value
+        ARG_SLOT invokeRetVal{};
+
+        // Invoke the method
+        forwardCallToInvoke.CallWithValueTypes_RetArgSlot(invokeArgs, &invokeRetVal, sizeof(invokeRetVal));
+
+        // Ensure all outs and return values are moved back to the current callsite
+        gc.RetVal = ArgSlotToObj(invokeRetVal);
```

Comment:
You can use `CallWithValueTypes_RetOBJECTREF` to make this simpler.

Reply:
Yay!